src/main/java/org/tridiots/ipc/Server.java

```diff
@@ -1,61 +1,100 @@
+        channel.configureBlocking(false);
+        channel.register(selector, SelectionKey.OP_READ);
+
+        logger.debug("accept a connection from {}.", channel.socket().getInetAddress());
+    }
+
+    private void doRead(SelectionKey key) throws IOException, ClassNotFoundException {
+        SocketChannel channel = (SocketChannel) key.channel();
+        logger.debug("read from connection from {}.", channel.socket().getInetAddress().getHostName());
+        Param param = (Param) SocketObjectUtil.receiveObject(channel);
+
+        if (param == null) {
+            // the channel will always readable if we don't close
+            channel.close();
+            return;
+        }
+        Object result = call(param);
+
+        key.attach(result);
+        key.interestOps(SelectionKey.OP_WRITE);
+    }
+
+    private void doWrite(SelectionKey key) throws IOException, ClassNotFoundException {
+        Object result =  key.attachment();
+        if (result == null) return;
+
+        SocketChannel channel = (SocketChannel) key.channel();
+        SocketObjectUtil.sendObject(channel, result);
+        key.interestOps(SelectionKey.OP_READ);
+
+        logger.debug("write to connection {}.", channel.socket().getInetAddress().getHostName());
     }
 
     @Override
     public void run() {
         while (running) {
-            Socket socket = null;
             try {
-                socket = serverSocket.accept();
-                System.out.println("get connection from " + socket.getInetAddress().getHostName());
-            } catch (IOException e) {
-                e.printStackTrace();
-            }
+                if (selector.select() == 0) continue;
 
-            try {
-                final ObjectOutputStream oos = new ObjectOutputStream(socket.getOutputStream());
-                final ObjectInputStream ois = new ObjectInputStream(socket.getInputStream());
-                pool.submit(() -> {
-                    while (running) {
-                        try {
-                            Param param = (Param)ois.readObject();
-                            Object result = call(param);
-                            oos.writeObject(result);
-                        } catch (ClassNotFoundException e) {
-                            e.printStackTrace();
-                        } catch (EOFException e) {
-                            break; // Client closed.
-                        } catch (IOException e) {
-                            e.printStackTrace();
-                        }
+                Iterator<SelectionKey> iter = selector.selectedKeys().iterator();
+                while (iter.hasNext()) {
+                    SelectionKey key  = iter.next();
+                    iter.remove();
+
+                    if (key.isAcceptable()) {
+                        doAccept(key);
+                    } else if (key.isReadable()) {
+                        doRead(key);
+                    } else if (key.isWritable()) {
```

Comment:
now `Server` is a real single thread, all the network IO and `namenode` functions will be run in this thread, and it will be bottleneck easily.